lib/winapi/filemanip.c

```diff
@@ -241,6 +241,138 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
     }
 }
 
+BOOL CopyFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName, BOOL bFailIfExists)
+{
+    NTSTATUS status;
+    HANDLE sourceHandle;
+    HANDLE targetHandle = INVALID_HANDLE_VALUE;
+    ANSI_STRING sourcePath;
+    ANSI_STRING targetPath;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_NETWORK_OPEN_INFORMATION networkOpenInformation;
+    LPVOID readBuffer = NULL;
+    SIZE_T readBufferRegionSize = 0x1000;
+    DWORD bytesRead;
+
+    RtlInitAnsiString(&sourcePath, lpExistingFileName);
+    InitializeObjectAttributes(&objectAttributes, &sourcePath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(
+            &sourceHandle,
+            FILE_GENERIC_READ,
+            &objectAttributes,
+            &ioStatusBlock,
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+        sourceHandle,
+        &ioStatusBlock,
+        &networkOpenInformation,
+        sizeof(networkOpenInformation),
+        FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    //
+    // NOTE: FATX does not have ".." entries in directory tables, so it is not
+    // possible to overlap a file via "subdir/../file".
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtCreateFile(
+            &targetHandle,
+            FILE_GENERIC_WRITE,
+            &objectAttributes,
+            &ioStatusBlock,
+            &networkOpenInformation.AllocationSize,
+            networkOpenInformation.FileAttributes,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
```

Comment:
This is a bug: The file should be opened with `0` as the sharing flags, not `FILE_SHARE_READ | FILE_SHARE_WRITE`, or another thread could open the same file and mess with it while we're copying. Using `0` will also make the call fail if we're trying to open the source file again.

Reply:
It's fine like this.